### ISSUE TITLE
On a wip branch, there are some subtle soundness issues with Lazy

### ISSUE THREAD

**Issue by matklad** [BurntSushi/regex-automata#30]
Just couldn't resist taking a look at a couple of safety issues that have burnt me in the past in this area :-)

This type

https://github.com/BurntSushi/regex-automata/blob/1d8f4d67e57bba5146627d5f8625db2abb925175/src/util/lazy.rs#L110

has two problems:

First, it misses `unsafe impl<T: Sync + Send, F: Sync> Sync for Lazy<T, F>` (the inferred impl misses + Send). This can be weaponized to turn Lazy into a channel for sending Sync+!Send types across threads: 

```rust
fn sync_not_send<T: Sync + Default>() {
    let lazy = Lazy::new(move || T::default());
    std::thread::scope(|scope| {
        scope.spawn(|| {
            lazy.get(); // We create T in this thread
        });
    });
    // And drop in this thread.
    drop(lazy);
    // So we have send a !Send type over threads. (with some more legwork, its
    // possible to even sneak the value out of drop through thread local)
}
```

Second, the type owns `T` (it can drop it), but this isn't visible to compiler. That is, `PhantomData<Box<T>>` is missing. In principle, this should be weaponizable by subverting dropcheck. If the compiler believes that dropping `Lazy` can't drop `T`, then it might allow code where the T in lazy has some dangling pointers when the Lazy is dropped, something along the lines of 

```rust
fn no_dropchk() {
    struct Sneaky<'a>(&'a str);
    impl<'a> Drop for Sneaky<'a> {
        fn drop(&mut self) {
            println!(".0 = {}", self.0);
        }
    }

    let lazy;
    let s = " hello".to_string();
    lazy = Lazy::new(|| Sneaky(s.as_str()));
}
```

No, this exact code compiles, because, although compiler doesn't see that we can drop `T`, it sees that we can drop `F`, and those share the problematic lifetime. I am maybe 70% confident that, because of this, there's no way too weaponize it, but still adding phantom data would seem like a good idea. 

(newer versions of OnceCell crate also have this version of synchronization primitive: https://docs.rs/once_cell/latest/once_cell/race/struct.OnceBox.html. We don't have a spnning no-alloc version though, as I somewhat firmly believe that, for all code that _could_ spin lock wihtout yielding to OS/blocking interrupts/etc, the caller should explicitly type out `.spin` somewhere)

**Comment by BurntSushi** [comment]
Ahhhhhh yesss! My secret plan to have someone review this code worked. :D

Could I convince you to review `Pool` too? It's not quite the same as `Lazy`, but it is similarish. And I just caught a soundness hole in it a few weeks ago. The last three tests in `src/util/pool.rs` tell that story.

> We don't have a spnning no-alloc version though, as I somewhat firmly believe that, for all code that could spin lock wihtout yielding to OS/blocking interrupts/etc, the caller should explicitly type out .spin somewhere)

Yeah, I've read your blog post a few times and I find it pretty compelling. _Buuuuuuuuuuut_, I don't really know what else to do. For `Lazy`, the core-only functionality is there so that folks in no-std no-alloc environments have an easier path. Namely, one of the main use cases of `regex-automata` is the ability to serialize a DFA, and then deserialize it in a much more constrained environment (no-std _and_ no-alloc). I actually don't have a ton of experience in said environments, so I don't quite know just how useful having something like a `Lazy` type is. But, it certainly makes things like `regex-cli generate serialize` easier because it just write out Rust code that uses library types from `regex-automata` to make a DFA globally available with zero fuss. But, I don't feel too bad about this, because nobody has to use `Lazy`. It isn't used inside the regex internals anywhere. (Currently, anyway. There are use cases for it, but for a variety of reasons unrelated to `Lazy`, I haven't been able to make it work as good as I've wanted.)

With that said, I'm _also_ using a spinlock in `Pool`, and that gets used when only `std` is disabled. So if you have `alloc` but not `std`, then you get a `Pool` with a spinlock. (`Pool` is unavailable in no-std and no-alloc environments because it's pretty heavily tied to the idea of allocating something.) In practice where this is going to show up is when folks disable the `std` feature in the `regex` crate (which folks have been asking for for a while). So when they do that, a call to say, `Regex::find`, will ask the pool for some memory to run a search. And that's going to hit the spinlock.

Now, in regex-automata, there is an [escape hatch](https://burntsushi.net/stuff/tmp-do-not-link-me/regex-automata/regex_automata/meta/struct.Regex.html#synchronization-and-cloning) for all of this. You can choose to use a lower level API that lets you pass your own mutable memory, which means the regex doesn't need to talk to a pool at all. But, `regex` proper doesn't have those APIs, and I don't really _want_ it to have those APIs because it fattens up an already big API for a very niche use case. Maybe eventually I'll cave on that point, but at least at present, I don't feel strongly enough about spinlocks to choose a different point in the design space as I currently conceive it.

I had also been hoping to get to the section on spinlocks in @m-ou-se's book (pinging you on the off chance you want to chime in here) to see if there was any extra insight to be gathered there.

Mara: big picture thing here is that, at some point in the next few months, the `regex` crate will provide a way to build without `std`. When it does this, it still needs to use a synchronization primitive to get some mutable scratch space in order to execute a search. Basically, think of it as a `Mutex<Vec<SomeRegexMemory>>`. It plucks something out at the start of a search, does some reading & writing to it, and then stuffs it back in when it's done so that another search can reuse it. Now with `std` enabled, it does actually use a `Mutex` (with some other tricks). But when you disable `std`, currently, my plan is to have it use a spinlock because I don't really know what else to do _if_ I want to keep providing the nice high level `Regex::find` APIs (and I do).

**Comment by BurntSushi** [comment]
@matklad Also, I've pushed up fixes to your feedback. Thank you. I've also added big warnings to `Pool` and `Lazy` (with a linkt to your blog post).

**Comment by BurntSushi** [comment]
This is the current code that implements the spinlock mentioned above for a `Pool`:

https://github.com/BurntSushi/regex-automata/blob/392b1569c1459a11a151b2e31756afdf01723c3e/src/util/pool.rs#L729-L815

**Comment by m-ou-se** [comment]
> Mara: big picture thing here is that, at some point in the next few months, the `regex` crate will provide a way to build without `std`. When it does this, it still needs to use a synchronization primitive to get some mutable scratch space in order to execute a search. Basically, think of it as a `Mutex<Vec<SomeRegexMemory>>`. It plucks something out at the start of a search, does some reading & writing to it, and then stuffs it back in when it's done so that another search can reuse it. Now with `std` enabled, it does actually use a `Mutex` (with some other tricks). But when you disable `std`, currently, my plan is to have it use a spinlock because I don't really know what else to do _if_ I want to keep providing the nice high level `Regex::find` APIs (and I do).

If that Vec is just used as a stack, to which you only push and pop only one element at a time that will only be used by one thread, it might be simple to do that lock free. :)

**Comment by matklad** [comment]
My general stance on spinlocks is driven primarily by aesthetic reasons: it's no that I debugged some curious latency spikes in a server, it's just that I understand the mechanics of how spin locks could lead to issues, and that shifts me to "it's not up to the library to make this choice". Maybe in practice it is actually fine! But maybe not: [here's someone hitting an actual deadlock in the crossbeam](https://github.com/crossbeam-rs/crossbeam/issues/821). I guess that issue can be used to create a litmus test here:

* the user is using a `no_std` regex in user-land code (eg, they are using some library which doesn't need and doesn't have `std` version, and that library depends on `Regex` with `default-features = false`).
* the application uses two threads, pinned to the same CPU, with one thread being higher priority
* two threads concurrently try to do a regex search, and, with an unfortunate interleaving, the result is a deadlock

That does seem quite unlikely, but also is not impossible, and the result (deadlock) is almost as bad as it can go. Is it ok for a library to potentially open the used to this scenario? On the aesthetic ground definitely not! On practical grounds, it might actually be yes? 

---

One thing I've realized here is that no_std+alloc is in some sense an incoherent world. If you have a global allocator, you _must_ have a proper implementation of Mutex to protect it (that impl can be a no-op for threadless environments). This is somewhat ironic in this situation --- you are building a globabl memory pool, a specialized version of a global allocator, you need the same kinds of tools to build that, the tools are actually in your address space (as they are used by the global allocator), but you can't reach them!

---

I think for pool what Mara suggests makes the most sense. Lock-free pool + racy lazy should be totally fine. I am not sure what to do with no_std no_alloc lazy... I see that it is used to intialized an &[u32]. As the initialization is deterministic, I am wondering if it would be correct to do something like:

```rust
if !initialized.load(Acquire) {
  for byte in bytes.iter() { byte.store(deterministic_value, Relaxed) }
  initialized.store(true, Release)
}
for byte in bytes.iter() {
  byte.load(NonAtomic);
}
```

That is, threads race to overwrite the byte array, but, because they are guaranteed to write the same value, it is actually ok? My reading of C++ memory model tells me that no, this won't be ok. On the high level, I believe the memory model doesn't have anything specific about "same value", so reasoning relying on that would we wrong. More specifically, as we ultimately want to do non-atomic reads, then our atomic relaxed writes do nothing. And, if I remove that relaxed store, it seems clear that this is UB --- you have non-atomic read and write racing with each other, and that's definitely UB, even if the write overwites the same value. 

Curiously though, if you also make sure that reads are atomic relaxed, that should be fine! (but then you'd have to return a `&[AtomicU32]`).

---

```rust
pub static {name}: Lazy<DFA<&'static [u32]>> = Lazy::new(|| {{
    {deserialize}
}});
```

I'd maybe write this as

```rust
pub fn {name}() DFA<&'static [u32]> {
    static LAZY = Lazy<DFA<&'static [u32]>> = Lazy::new(|| {{
        {deserialize}
    }});
    &*LAZY
}
```

to not expose specific synchronization in the API.

---

As to what to do here in general, I think there are these choices:

* Figuring out what the users actually _want_ when they ask for "no_std regex". It might be either of the two:
  * I want to match a string with a regex on my embedded device
  * I want to re-use the _api_ of the regex crate as is, such that I get access to oodles of regex-using crates for free

  My personal stance here is that no_std is a different, resource constrained environment and so in the general case it just needs to have a different set of APIs. This makes composing libraries hard, but, if you care about resource-constrained environments, you probably care about zero-cost abstractions more than you care about seamless composition and would be fine with writing some amount of glue code yourself.
* If there's a requirement to keep the API, then you can either silently default to spinlock, or you can make the whole thing pluggable from the outside. I _think_ https://docs.rs/critical-section/latest/critical_section/ and https://docs.rs/getrandom/latest/getrandom/#custom-implementations do this, and the trick boils down to, at the lower level, `extern "C" { fn hook(); }` declaration in the upstream crate which is satisfied by the downstream crate and ultimately resolved by the linker. 
* You can also make this someone else's problem! In particular, `lazy_static` I believe falls back to spinlocks for no_std, so, by (optionally) depending on it, you absolve your own crate from questionable code. ["If you just want spinlocks, use lazy_static with all of the caveats"](https://github.com/matklad/once_cell/issues/223#issuecomment-1442304743) I think is a fine strategy.

**Comment by BurntSushi** [comment]
@m-ou-se thank you! The reason why i hadn't give the approach of a lock free structure was indeed because of the ABA problem. It looks like you get around that by still using a spin lock. Does that still fall afoul of the problems with spin locks in general though? (Yours has much smaller scope than mine, of course.)

**Comment by m-ou-se** [comment]
A spinlock isn't generally bad. A regular `Mutex` also spins for quite a while before it blocks. Using a spinlock when it can sometimes spin for a long time is where it starts becoming a bad idea. In this case, it is only locked between the compare-exchange and the store that immediately follows, not while performing any allocations or syscalls etc. (An `Arc` does exactly the same in `Arc::downgrade()`.)

**Comment by m-ou-se** [comment]
(But as usual, only a benchmark will provide real answers. ^^)

**Comment by m-ou-se** [comment]
> It looks like you get around that by still using a spin lock.

(I would say "by using spinning", not "by using a spin lock". A `SpinLock<Vec<T>>` would need to be locked on every access. My implementation only briefly 'locks' for popping, not for pushing or allocating (or resizing a Vec).)

**Comment by matklad** [comment]
> I guess what I had in mind is whether @matklad's deadlock scenario can happen in your pool.

Yeah, I believe it could happen if:

* there are two threads pinned to a single core
* the OS facilities are used to enforce "thread 1 has higher priority than thread 2, so, if both are runnable, 1 should always receive higher priority"

https://github.com/matklad/spin-of-death/blob/master/src/main.rs (from my post) might be a good starting point to create the above situation.

>A spinlock isn't generally bad.

I think it's important to make a distinction between "spinning a bit before syscall" (awesome perf optimization!) and an actual spinlock, where we burn the CPU, blocked on the progress of some other thread (can create a deadlock if thread priorities are unfair).

**Comment by m-ou-se** [comment]
Note that `alloc::sync::Arc`'s `downgrade()` and `get_mut()` have the exact same problem in those situations. I'll leave it up to you to decide if that means "`Arc` has a bug" or "those situations are apparently unsupported if you use `alloc`" and what that means for `regex`. :)

**Comment by matklad** [comment]
Got this to deadlock without signals, just using realtime thread priorities: 

https://github.com/matklad/spin-of-death/commit/7ed9770913ddeda6188c752bab47fe6c142422e9

But yeah, it indeed seems like `Arc::downgrade` can deadlock in a similar way if a different lower priority thread does `get_mut` in a bad moment.

**Comment by BurntSushi** [comment]
Aye, thanks for this everyone! I think I'm just going to swap out my alloc-only implementation of `Pool` with @m-ou-se's from above. I think I'm comfortable being in a similar spot as `Arc::downgrade` in exchange for keeping the nicer higher level API. And I'll add warnings to the docs. I also feel okay about this because if someone is concerned about it, they _will_ have an escape route by using lower level APIs that don't make use of a `Pool` at all.
